Adding global-timeout, individual command timeout, log files collection

fk410167 opened the pull request:
and some other enhancements to techsupport

<!--
Please make sure you've read and understood our contributing guidelines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "closes #xxxx",
"fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
issue when the PR is merged.

If you are adding/modifying/removing any command or utility script, please also
make sure to add/modify/remove any unit tests from the tests
directory as appropriate.

If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
subcommand, or you are adding a new subcommand, please make sure you also
update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
your changes.

Please provide the following information:
-->

**- What I did**
Following is the brief description of the changes,
- Adding a ‘--silent’ option to ‘show techsupport’ command. Various  tar/untar, addition and removal logs appear on the console by default. This option would disable above logs.
- Adding global and per-command timeouts. This would provide more user control on ‘show techsupport’ CLI.
- Adding time profiling information for the commands in techsupport. Time profiling information would be part of the tarball and helps to analyse the time consumption per command.
- Sometimes ‘syncd’ docker is down and bcmshell is unavailable. In such cases all the bcmcmd commands would timeout and result in tremendous increase in the total techsupport collection time. We provided an option to skip rest of the bcmcmd commands once one command times out.
- Added ‘show services’, ‘show reboot-cause’ and various BGP, BFD, bcm shell and other commands
- Optimised the /var/log files collection. If the number of files are large in /var/log folder, it takes a long time to add each individually to the tarball. If the folder is tar'ed at once, the time taken reduces significantly. 
- Following error was observed while tar'ing softlinks inside .etc folder. 
  ** Tar append operation failed. Aborting for safety. **
  This issue was due to softlinks present at /etc folder where the destination file is absent. Fixed this issue by deleting such softlinks before adding them to the tarball. 

**- How I did it**
- Added new options to the CLICK command 'show techsupport'
- Modified the 'generate_dump' script to accomodate other changes

**- How to verify it**
Here are some outputs,
root@sonic:/home/admin# show techsupport --silent
Techsupport is running with silent option. This command might take a long time.

HW Mgmt dump script /usr/bin/hw-management-generate-dump.sh does not exist
/var/dump/sonic_dump_sonic_20201117_161246.tar.gz
root@sonic:/home/admin#
root@sonic:~# show techsupport -h
Usage: show techsupport [OPTIONS]

  Gather information for troubleshooting

Options:
  --since TEXT                  Collect logs and core files since given date
  -g, --global-timeout INTEGER  Global timeout for techsupport in minutes.
                                Default 30 mins
  -c, --cmd-timeout INTEGER     Command timeout for techsupport in minutes.
                                Default 5 mins
  --verbose                     Enable verbose output
  --silent                      Run techsupport in silent mode
  -?, -h, --help                Show this message and exit.
root@sonic:~#

**- Previous command output (if the output of a command-line utility has changed)**
- Previous command "show techsupport" works as is
**- New command output (if the output of a command-line utility has changed)**

fk410167 commented:
retest default please

fk410167 commented:
retest this please

fk410167 commented:
retest this please

smaheshm commented:
Thanks for enhancing the tech support collection. One suggestion I have is next time PRs like this can be split into multiple PRs with each PR having a specific fix or enhancement. This makes it easier to revert a PR when there's an issue, also the PR looks cleaner and easier on the reviewer.

smaheshm commented:
Have you raised or going to raise a PR for tech support unit tests
https://github.com/Azure/sonic-mgmt/tree/master/tests/show_techsupport

smaheshm commented on scripts/generate_dump:
```diff
@@ -25,15 +34,87 @@ DO_COMPRESS=true
+    then
+        if $NOOP; then
+            echo "${timeout_cmd} $cmd &> '${filepath}'"
+        else
+            eval "${timeout_cmd} $cmd" &> "${filepath}"
+            ret=$?
+            if [ $ret -ne 0 ]; then
+                if [ $ret -eq 124 ]; then
+                    echo "Command: $cmd timedout after ${TIMEOUT_MIN} minutes."
+                else
+                    grep "polling socket timeout: Success" ${filepath} &>/dev/null
+                    if [ $? -eq 0 ]; then
+                        echo "bcmcmd command timeout. Setting SKIP_BCMCMD to true ..."
+                        SKIP_BCMCMD=1
+                    fi
+                fi
+            fi
+        fi
+        gzip ${filepath} 2>/dev/null
+        tarpath="${tarpath}.gz"
+        filepath="${filepath}.gz"
+    else
+        if $NOOP; then
+            echo "${timeout_cmd} $cmd &> '$filepath'"
+        else
+            eval "${timeout_cmd} $cmd" &> "$filepath"
+            ret=$?
+            if [ $ret -ne 0 ]; then
+                if [ $ret -eq 124 ]; then
+                    echo "Command: $cmd timedout after ${TIMEOUT_MIN} minutes."
+                else
+                    grep "polling socket timeout: Success" ${filepath} &>/dev/null
+                    if [ $? -eq 0 ]; then
+                        echo "bcmcmd command timeout. Setting SKIP_BCMCMD to true ..."
+                        SKIP_BCMCMD=1
+                    fi
+                fi
+            fi
+        fi
+    fi
```
The only difference between the 'zip' and 'no zip' option seems to be the 3 statements. Can you refactor the code to reduce duplicate statements.

smaheshm replied:
how about:
```
if $do_gzip
        gzip ${filepath} 2>/dev/null
        tarpath="${tarpath}.gz"
        filepath="${filepath}.gz"
fi
```

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -25,15 +34,87 @@ DO_COMPRESS=true
+    if $do_gzip
+    then
+        if $NOOP; then
+            echo "${timeout_cmd} $cmd &> '${filepath}'"
```
is there a need to provide "timeout_cmd" with 'NOOP' option.

smaheshm replied:
never mind, this is just echoing the cmd.. this is correct.

smaheshm commented on scripts/generate_dump:
```diff
@@ -98,10 +190,10 @@ save_cmd() {
 save_cmd_all_ns() {
-    echo $1
-    echo $2
+    #echo $1
```
You can remove the commented echo statements here.

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -186,6 +278,24 @@ save_ip() {
 }
 
+###############################################################################
+# Runs an bridge command and saves its output to the incrementally built tar.
```
"Runs a bridge"

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -698,16 +1003,25 @@ main() {
     save_ip_info
+
+    save_bridge "fdb show" "fdb"
+    save_bridge "vlan show" "vlan"
```
How about rename API to 'save_bridge_info' and move these two to that API. In future all l2 info collection can be put in there.

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -252,6 +372,23 @@ save_nat_info() {
+# Returns:
+#  None
+###############################################################################
+save_bfd_info() {
```
👍

smaheshm commented on scripts/generate_dump:
```diff
@@ -536,17 +711,59 @@ collect_mellanox() {
-    save_cmd "bcmcmd -t5 soc" "broadcom.soc"
-    save_cmd "bcmcmd -t5 ps" "broadcom.ps"
+    local platform="$(/usr/local/bin/sonic-cfggen -H -v DEVICE_METADATA.localhost.platform)"
+    local hwsku=$(head -n 1 /usr/share/sonic/device/${platform}/default_sku | cut -d ' ' -f 1)
```
use 'show platform summary --json' output. Checkout 'get_asic_count()'

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -536,17 +711,59 @@ collect_mellanox() {
-    save_cmd "cat /proc/bcm/knet/stats" "broadcom.knet.stats"
-    save_cmd "bcmcmd \"l3 nat_ingress show\"" "broadcom.nat.ingress"
-    save_cmd "bcmcmd \"l3 nat_egress show\"" "broadcom.nat.egress"
+    save_bcmcmd "bcmcmd \"l3 nat_ingress show\"" "broadcom.nat.ingress"
+    save_bcmcmd "bcmcmd \"l3 nat_egress show\"" "broadcom.nat.egress"
+    save_bcmcmd "bcmcmd \"ipmc table show\"" "broadcom.ipmc"
+    save_bcmcmd "bcmcmd \"multicast show\"" "broadcom.multicast"
+    save_bcmcmd "bcmcmd \"conf show\"" "conf.summary"
+    save_bcmcmd "bcmcmd \"fp show\"" "fp.summary"
+    save_bcmcmd "bcmcmd \"pvlan show\"" "pvlan.summary"
+    save_bcmcmd "bcmcmd \"l2 show\"" "l2.summary"
+    save_bcmcmd "bcmcmd \"l3 intf show\"" "l3.intf.summary"
+    save_bcmcmd "bcmcmd \"l3 defip show\"" "l3.defip.summary"
+    save_bcmcmd "bcmcmd \"l3 l3table show\"" "l3.l3table.summary"
+    save_bcmcmd "bcmcmd \"l3 egress show\"" "l3.egress.summary"
+    save_bcmcmd "bcmcmd \"l3 ecmp egress show\"" "l3.ecmp.egress.summary"
+    save_bcmcmd "bcmcmd \"l3 multipath show\"" "l3.multipath.summary"
+    save_bcmcmd "bcmcmd \"l3 ip6host show\"" "l3.ip6host.summary"
+    save_bcmcmd "bcmcmd \"l3 ip6route show\"" "l3.ip6route.summary"
+    save_bcmcmd "bcmcmd \"mc show\"" "multicast.summary"
+    save_bcmcmd "bcmcmd \"cstat *\"" "cstat.summary"
+    save_bcmcmd "bcmcmd \"mirror show\"" "mirror.summary"
+    save_bcmcmd "bcmcmd \"mirror dest show\"" "mirror.dest.summary"
+    save_bcmcmd "bcmcmd \"port *\"" "port.summary"
+    save_bcmcmd "bcmcmd \"d chg my_station_tcam\"" "mystation.tcam.summary"
```
all the 'bcmcmd' commands must be run for all ASICs on multi ASIC platforms. The format is 
```
bcmcmd -n <asic ID> <options..>
```
Please also update the ones that already existed and not updated.

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -536,17 +711,59 @@ collect_mellanox() {
+    if [ -d /usr/share/sonic/device/${platform}/${hwsku} ]; then
+        # copy all the files in the HWSKU directory
+        pushd /usr/share/sonic/device/${platform}/${hwsku} > /dev/null
+        for file in $(find . -maxdepth 1 -type f); do
```
change maxdept to '2' to include files for multi ASIC platforms as their file location is under ASIC ID subdir.

/usr/share/sonic/device/${platform}/${hwsku}/<asicID>

fk410167 replied:
Done

fk410167 replied:
done

smaheshm commented on scripts/generate_dump:
```diff
@@ -536,17 +711,59 @@ collect_mellanox() {
+        done
+        popd > /dev/null
+        # config.bcm - copy the one with chip common properties merged
+        for file in $(find /var/run/docker-syncd -type f -name "*.bcm"); do
```
Please include support for multi ASIC platforms, where the syncd docker are
/var/run/docker-syncd0
/var/run/docker-syncd1
.
.

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -642,6 +892,40 @@ get_asic_count() {
+
+    save_cmd "echo $counter_t" "date.counter_$idx"
+    save_cmd "show interface counters" "interface.counters_$idx"
+    save_cmd "show queue counters" "queue.counters_$idx"
```
use "save_cmd_all_ns" for "show queue counters", so that it works on multi ASIC platform

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -642,6 +892,40 @@ get_asic_count() {
+            save_cmd "cat /proc/bcm/knet/rx_drop" "broadcom.knet_queue.counters_$idx"
+        fi
+    fi
+    save_cmd "netstat -i" "netstat.counters_$idx"
+    save_cmd "ifconfig -a" "ifconfig.counters_$idx"
```
use "save_cmd_all_ns" for these two commands

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -698,16 +1003,25 @@ main() {
-    save_cmd "show interface counters -d all" "interface.counters"
     save_cmd "show interface transceiver presence" "interface.xcvrs.presence"
     save_cmd "show interface transceiver eeprom --dom" "interface.xcvrs.eeprom"
+    save_cmd "show ip interface" "ip.interface"
```
use "save_cmd_all_ns" for "show ip interface".

fk410167 replied:
Done

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -698,16 +1003,25 @@ main() {
+    save_cmd "show ip interface" "ip.interface"
 
     save_cmd "lldpctl" "lldpctl"
+    save_cmd "docker exec -it lldp lldpcli show statistics" "lldp.statistics"
+    save_cmd "docker logs bgp" "docker.bgp.log"
+    save_cmd "docker logs swss" "docker.swss.log"
```
lldp, bgp, swss services in multi ASIC platform have asic ID appended in the end.
bgp<asicID>, swss<asicID>, lldp<asicID>
eg. bgp0, bgp1...

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -756,8 +1080,18 @@ main() {
+    end_t=$(date +%s%3N)
+    echo "[ TAR /etc Files ] : $(($end_t-$start_t)) msec" >> $TECHSUPPORT_TIME_INFO
 
+    if [ "$asic" = "broadcom" ]; then
+       sudo docker exec -i syncd touch /var/log/diagrun.log
+       sudo docker exec -i syncd touch /var/log/bcm_diag_post
+
+       sudo docker cp syncd:/var/log/diagrun.log /var/log/diagrun.log
+       sudo docker cp syncd:/var/log/bcm_diag_post /var/log/bcm_diag_post
```
same as above, services are appended with asicID.

fk410167 replied:
Done

smaheshm reviewed (changes requested):
I'd suggest to break this into multiple PRs, up to you.

fk410167 reviewed (commented):
- As most of the review comments are regarding the multi_asic, I would like to keep all this in single PR.
- I have done manual testing
- I am adding some tests in sonic-mgmt for the changes

smaheshm commented on scripts/generate_dump:
```diff
@@ -25,15 +34,115 @@ DO_COMPRESS=true
+# Returns:
+#  None
+###############################################################################
+save_bcmcmd_all_ns() {
```
👍

smaheshm commented on scripts/generate_dump:
```diff
@@ -186,6 +303,39 @@ save_ip() {
+save_bridge_info() {
+    save_bridge "fdb show" "fdb"
+    save_bridge "vlan show" "vlan"
+
```
space can be removed

fk410167 replied:
done

smaheshm commented:
Thanks for making the changes. Looks good. 
My main concern is the unit test for existing test cases. Would like to see the output of pytest. 

Since you added new features, in case you are still working on writing unit tests for the new features, would like to see the output. For example.. the output of "global timeout" case and the "command timeout" case. I'm sure you've tested it, just 'copy/paste' the output.

Regarding "I am adding some tests in sonic-mgmt for the changes" -- have you raised a PR or are working on it.

Adding @qiluo-msft for another pair of eyes.

Rest looks good to me.

smaheshm commented on show/main.py:
```diff
@@ -1224,12 +1224,23 @@ def users(verbose):
 
 @cli.command()
 @click.option('--since', required=False, help="Collect logs and core files since given date")
+@click.option('-g', '--global-timeout', default=30, type=int, help="Global timeout for techsupport in minutes. Default 30 mins")
```
for the next two lines, how about:
```
@click.option('
    -g', '--global-timeout', default=30, type=int,
    help="Global timeout for techsupport in minutes. Default 30 mins"
)
```

fk410167 replied:
done

fk410167 replied:
I modified the help string to keep the line width <= 120 chars

smaheshm commented on scripts/generate_dump:
```diff
@@ -756,8 +1145,29 @@ main() {
+        fi
+    fi
     save_log_files
+	save_warmboot_files
```
extra space?

fk410167 replied:
done

smaheshm commented on show/main.py:
```diff
@@ -1224,12 +1224,23 @@ def users(verbose):
+
+    if silent:
+        cmd += " generate_dump"
+        click.echo("Techsupport is running with silent option. This command might take a long time.\n")
```
is "\n" required here?

fk410167 replied:
Done

smaheshm commented on scripts/generate_dump:
```diff
@@ -680,13 +1029,27 @@ main() {
+    # Save reboot cause information
+    save_cmd "show reboot-cause" reboot.cause
+
+
```
remove extra empty line?

fk410167 replied:
done

smaheshm commented on scripts/generate_dump:
```diff
@@ -642,6 +950,40 @@ get_asic_count() {
+    save_cmd_all_ns "netstat -i" "netstat.counters_$idx"
+    save_cmd_all_ns "ifconfig -a" "ifconfig.counters_$idx"
+}
+
```
remove extra empty line?

fk410167 replied:
done

fk410167 commented:
Still working on sonic-mgmt test PR. I will add it later. 

I ran some manual tests locally on the setup. Here are the logs,

root@sonic:/home/admin# show techsupport --silent
Techsupport is running with silent option. This command might take a long time.
tar: sonic_dump_sonic_20201216_164906/etc/sonic/frr/bgpd_client_sock: socket ignored
tar: sonic_dump_sonic_20201216_164906/etc/sonic/old_config/frr/bgpd_client_sock: socket ignored
/var/dump/sonic_dump_sonic_20201216_164906.tar.gz
root@sonic:/home/admin#
root@sonic:/home/admin#
root@sonic:/home/admin#
root@sonic:/home/admin# time show techsupport -g 1

main
mkdir: created directory '/var/dump/sonic_dump_sonic_20201216_165622'
'/var/dump/sonic_dump_sonic_20201216_165622/generate_dump' -> '/usr/local/bin/generate_dump'
sonic_dump_sonic_20201216_165622/
sonic_dump_sonic_20201216_165622/generate_dump
mkdir: created directory '/var/dump/sonic_dump_sonic_20201216_165622/proc'
'/proc/buddyinfo' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/buddyinfo'
'/proc/cmdline' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/cmdline'
'/proc/consoles' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/consoles'
'/proc/cpuinfo' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/cpuinfo'
'/proc/devices' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/devices'
'/proc/diskstats' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/diskstats'
'/proc/dma' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/dma'
'/proc/interrupts' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/interrupts'
'/proc/iomem' -> '/var/dump/sonic_dump_sonic_20201216_165622/proc/iomem'
...
...
... 
sonic_dump_sonic_20201216_165622/dump/interface.xcvrs.presence
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/interface.xcvrs.presence'
sonic_dump_sonic_20201216_165622/dump/interface.xcvrs.eeprom
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/interface.xcvrs.eeprom'
sonic_dump_sonic_20201216_165622/dump/ip.interface
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/ip.interface'
sonic_dump_sonic_20201216_165622/dump/lldpctl
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/lldpctl'
sonic_dump_sonic_20201216_165622/dump/lldp.statistics
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/lldp.statistics'
sonic_dump_sonic_20201216_165622/dump/docker.bgp.log
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/docker.bgp.log'
sonic_dump_sonic_20201216_165622/dump/docker.swss.log
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/docker.swss.log'
sonic_dump_sonic_20201216_165622/dump/ps.aux
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/ps.aux'
sonic_dump_sonic_20201216_165622/dump/top
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/top'
sonic_dump_sonic_20201216_165622/dump/free
removed '/var/dump/sonic_dump_sonic_20201216_165622/dump/free'

real	1m1.016s
user	0m31.997s
sys	0m9.279s
root@sonic:/home/admin#

fk410167 commented:
retest this please

lguohan commented on scripts/generate_dump:
```diff
@@ -537,17 +732,77 @@ collect_mellanox() {
-    save_cmd "cat /proc/bcm/knet/dstats" "broadcom.knet.dstats"
     save_cmd "cat /proc/bcm/knet/link" "broadcom.knet.link"
     save_cmd "cat /proc/bcm/knet/rate" "broadcom.knet.rate"
-    save_cmd "cat /proc/bcm/knet/stats" "broadcom.knet.stats"
```
not sure why this is removed.

fk410167 replied:
@Iguohan
It is not removed. These stats were moved under save_counter_snapshot() function. We need to take two snapshots of the counter stats to get the counter trend.

lguohan commented on scripts/generate_dump:
```diff
@@ -537,17 +732,77 @@ collect_mellanox() {
+
     save_cmd "cat /proc/bcm/knet/debug" "broadcom.knet.debug"
     save_cmd "cat /proc/bcm/knet/dma" "broadcom.knet.dma"
-    save_cmd "cat /proc/bcm/knet/dstats" "broadcom.knet.dstats"
```
not sure why this is removed

fk410167 replied:
It is not removed. These stats were moved under save_counter_snapshot() function. We need to take two snapshots of the counter stats to get the counter trend.

ben-gale commented:
Note that this PR does not contain sonic-utilities UT because we are extending a command that does not have such a test today. However we will fix this in a follow-up PR.

fk410167 commented:
retest this please

lguohan commented:
> Thanks for enhancing the tech support collection. One suggestion I have is next time PRs like this can be split into multiple PRs with each PR having a specific fix or enhancement. This makes it easier to revert a PR when there's an issue, also the PR looks cleaner and easier on the reviewer.

agree, that should be the rule. @ben-gale

fk410167 commented:
retest this please

fk410167 commented:
retest this please

smaheshm reviewed (approved):
Looks good. Please wait for one more approval.

ben-gale commented:
@lguohan - Can this one merge now pls?

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -26,15 +35,96 @@ DO_COMPRESS=true
+    local filepath="${LOGDIR}/$filename"
+    local do_gzip=${3:-false}
+    local tarpath="${BASE}/dump/$filename"
+    local timeout_cmd="timeout --foreground ${TIMEOUT_MIN}m"
```
>timeout [](start = 23, length = 7)

`bcmcmd` support timeout `-t`. Could you use that instead of `timeout` command?

fk410167 replied:
@qiluo-msft 
We can't predict the behaviour of 'bcmcmd' in case syncd is crashed/stuck/in bad shape. The change you are suggesting might require very extensive testing.

qiluo-msft replied:
This is the my only concern, others lgtm.

---
In reply to: [546269345](https://github.com/Azure/sonic-utilities/pull/1249#discussion_r546269345) [](ancestors = 546269345)

smaheshm replied:
'timeout' supersedes the '-t' option of 'bcmcmd'. Is the concern regarding 'bcmcmd' may end up in a bad state like leaving the socket open?

qiluo-msft replied:
`bcmcmd` is not well tested with signal termination. I am not sure.

---
In reply to: [550679331](https://github.com/Azure/sonic-utilities/pull/1249#discussion_r550679331) [](ancestors = 550679331,546269345)

fk410167 replied:
@qiluo-msft 
"Could you use that and maybe together with timeout command" --- The bcmcmd has a default timeout of 30 sec. If no argument is given using -t option, '-t 30' is automatically taken. So the outside 'timeout' command **is** being used together with bcmcmd timeout option.
```
root@sonic:/home/admin# bcmcmd
USAGE: bcmcmd [-f <sun_path>] -v <cmd>
  -v                         verbose mode
  -f                         domain socket filename, default /var/run/sswsyncd/sswsyncd.socket
  -t                         timeout in seconds, default 30
RETURN VALUE:
    0                        success
    5                        socket io error
   22                        invalid argument
   62                        timeout

root@sonic:/home/admin#
```

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -6,6 +6,15 @@
 
 set -u
 
+trap 'handle_signal' 2
```
>2 [](start = 21, length = 1)

Use signal name instead of magic number

fk410167 replied:
done

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -6,6 +6,15 @@
+
+handle_signal()
+{
+    echo "Generate Dump received interrupt"
```
>echo [](start = 4, length = 4)

echo to stderr?

fk410167 replied:
done

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -6,6 +6,15 @@
+handle_signal()
+{
+    echo "Generate Dump received interrupt"
+    $RM $V -rf $TARDIR
```
>$RM $V [](start = 4, length = 6)

These variables should be defined before this function.

fk410167 replied:
done

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -26,15 +35,96 @@ DO_COMPRESS=true
 CMD_PREFIX=
 SINCE_DATE="@0" # default is set to January 1, 1970 at 00:00:00 GMT
 REFERENCE_FILE=/tmp/reference
+TECHSUPPORT_TIME_INFO=/tmp/techsupport_time_info
```
>/tmp/techsupport_time_info [](start = 22, length = 26)

Better use `mktemp`

fk410167 replied:
done

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -26,15 +35,96 @@ DO_COMPRESS=true
+    if [[ ( "$NUM_ASICS" > 1 ) ]]; then
+        for (( i=0; i<$NUM_ASICS; i++ ))
+        do
+            local cmd="bcmcmd -n $i $1"
```
>bcmcmd -n [](start = 23, length = 9)

`bcmcmd` does not have `-n` option.

fk410167 replied:
The change was requested by @smaheshm.

bcmcmd has been modified to include the ASIC number. https://github.com/Azure/sonic-buildimage/pull/4926

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -135,7 +233,7 @@ get_vtysh_namespace() {
         ns=" -n  ${asic_id}"
     fi
-    echo "$ns"
+    #echo "$ns"
```
>#echo "$ns" [](start = 4, length = 11)

The function stdout is changed.

fk410167 replied:
Removed the comment

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -253,6 +393,23 @@ save_nat_info() {
+    save_vtysh "show bfd peers json" "frr.bfd.peers.json"
+    save_vtysh "show bfd peers counters json" "frr.bfd.peers.counters.json"
+
+}
```
Remove extra blank line

fk410167 replied:
done

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -335,13 +512,19 @@ save_proc() {
 ###############################################################################
 # Dumps all fields and values from given Redis DB.
 # Arguments:
+#  DB id: id of DB for sonic-db-cli
```
>DB id [](start = 3, length = 5)

The option has nothing to do with DB id. Why add?

fk410167 replied:
I changed the comments appropriately

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -199,10 +329,11 @@ save_ip() {
     local ns=$(get_vtysh_namespace $asic_id)
 
-    neighbor_list_v4=$(vtysh $ns -c "show ip bgp neighbors" | grep "BGP neighbor is" | awk -F '[, ]' '{print $4}')
+    neighbor_list_v4=$( ${timeout_cmd} vtysh $ns -c "show ip bgp neighbors" | grep "BGP neighbor is" | awk -F '[, ]' '{print $4}')
```
>  [](start = 23, length = 1)

Remove this extra blank

fk410167 replied:
Done

qiluo-msft commented on scripts/generate_dump:
```diff
@@ -578,23 +833,56 @@ save_log_files() {
+    echo "[ TAR /var/log Files ] : $(($end_t-$start_t)) msec" >> $TECHSUPPORT_TIME_INFO
+
+
     enable_logrotate
```
Remove extra blank line

fk410167 replied:
done

fk410167 commented:
retest this please

smaheshm commented:
retest this please

fk410167 commented:
retest this please

ben-gale commented:
@qiluo-msft - can this proceed to approve/merge now? All checks are passing and I believe review comments have been addressed.

ben-gale commented:
Pls merge
